### PR TITLE
Fix broken pipe error on docker layer creation

### PR DESCRIPTION
**Pull request by mickours**
This reverts commit f12346d49359784617a62731501627dc032419d8.

This commit was an optimization which was using pipes to read the layer tarball.
It was causing a 141 exit code ([means SIGPIPE for `tar`](https://stackoverflow.com/a/18219271/2165830)) while running Layer packing in the VM lauched for the runAsRoot command.
```
Packing layer...
[    3.755619] reboot: Power down
builder for '/nix/store/nzc2cpi2xh4whzq5h8dg7mgb5zns7993-docker-layer-mycontainer.drv' failed with exit code 141
```
This error happen consistently on multiple servers running Ubuntu 18.04 servers with Nix 20.09.

I've tested the patch and the error disappeared on all servers.

###### Motivation for this change

Fix this error by removing the optimization, so it won't happen again.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by utdemir** [comment]
Could you share a minimal example? Another way to solve it would be to figure out why we're getting `SIGPIPE`, and fix that issue instead.

Likely a command fails and the pipe is closed prematurely.

**Comment by pedrovelho** [comment]
> Could you share a minimal example? Another way to solve it would be to figure out why we're getting `SIGPIPE`, and fix that issue instead.
> 
> Likely a command fails and the pipe is closed prematurely.

I agree but SIGPIPE is rather a major security breech as is and should not be taken lightly : https://vigilance.fr/vulnerability/QEMU-denial-of-service-via-NBD-SIGPIPE-Signal-23103

**Comment by roberth** [comment]
> a major security breech

SIGPIPE is merely a normal way for unix systems to communicate that the process behind a pipe is gone. It is not security problem in and of itself. The issue you've linked is a flaw in QEMU, which is a distinct program that is not involved in this PR. Other than that, the Nix sandbox should not be used for running untrustworthy code because it has not been audited (as far as I know).

**Comment by mickours** [comment]
> Could you share a minimal example? Another way to solve it would be to figure out why we're getting `SIGPIPE`, and fix that issue instead.
> 
> Likely a command fails and the pipe is closed prematurely.

In fact both commands are working as expected because when run sequentially they finishe properly.
The issue is in the timing here. It seems that on my servers, tarsum finishes before tar and thus broke the pipe. This is probably due to a delay while tar is syncing the file on the file system (which is a VM and might be slow...).

An interesting post which explains that well: https://stackoverflow.com/a/38549483/2165830

Maybe a better solution would be to ignore the broken pipe using the `tee` option `-p` as mentioned in that post.
I'll try that and change the PR if it works.

**Comment by mickours** [comment]
Here is the new patch which only add `-p` option to the `tee` command (no revert). I've tested  it on my servers and works as expected. No more exit 141 errors.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
@utdemir looks like we need another approval before this can be merged.

**Review by jonringer — CHANGES_REQUESTED** [review]
please target `release-20.09`, `nixos-20.09` is a protected branch

To comply with [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes) please have the commit message name be of the format
```
<pkg-name>: <subject-line>
```
for more examples, please look at https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes

in your case, the commit message should be:
```
docker: Fix broken pipe error on layer creation
```

**Comment by jonringer** [comment]
When backporting changes, please follow https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#backporting-changes.

Namely, you should be doing `git cherry-pick -x <rev>` from a commit that has already landed in master. If the branches have diverged, you may alter the commit or add another commit to ensure that the package is able to still evaluate and build

**Comment by mickours** [comment]
Sorry @jonringer  I should have read the Contributing Guide before, thanks for the advice.

I've change the commit to what you've suggested and rebase my contribution on master. I'll make another PR for the backport into 20.09.

**Comment by roberth** [comment]
@mickours While figuring out what's going on with this PR, I noticed your commit does not have an author e-mail address that matches your GitHub account. You might want to fix that, either in https://github.com/settings/emails or in your git config and then `git commit --amend --reset-author`.

Other than that, it looks good, but to the person who will merge this, let's alter the branch name in the merge commit message to something sensible.

**Comment by mickours** [comment]
> @mickours While figuring out what's going on with this PR, I noticed your commit does not have an author e-mail address that matches your GitHub account. You might want to fix that, either in https://github.com/settings/emails or in your git config and then `git commit --amend --reset-author`.
> 
> Other than that, it looks good, but to the person who will merge this, let's alter the branch name in the merge commit message to something sensible.

Done!

**Comment by roberth** [comment]
Thank you!

**Pull request merged**
